pdq/abstractness.go

```diff
@@ -0,0 +1,19 @@
+package pdq
+
+import (
+	"fmt"
+)
+
+func CalcAbstractness(abstractElementsNum int, concreteElementsNum int) (float64, error) {
+	if abstractElementsNum < 0 || concreteElementsNum < 0 {
+		return -1, fmt.Errorf("abstractElementsNum %v, concreteElementsNum %v; values must be >= 0", abstractElementsNum, concreteElementsNum)
+	}
+
+	if concreteElementsNum == 0 {
+		return float64(abstractElementsNum), nil
+	}
+
+	abstractness := float64(abstractElementsNum) / float64(concreteElementsNum)
```

Comment:
if there are no concrete elements (concreteElementsNum == 0) but there are abstract elements, abstractness can be defined as 1 rather than float64(abstractElementsNum). If there are no abstract elements, then abstractness is equal to 0..or?

like 
```golang
	totalElements := abstractElementsNum + concreteElementsNum
	if totalElements == 0 {
		return 0, nil
	}

	abstractness := float64(abstractElementsNum) / float64(totalElements)
	return abstractness, nil
```

not sure check it

Reply:
Abstraction is measured in the range from 0 to 1, where 0 is the absence of abstraction, and 1, respectively, is complete abstraction.

But...

As I understand it, the formula assumes that the number of abstract elements is always <= the number of concrete ones (in practice, this is how it should be). This means that everything else is edge cases.

For example:

Valid cases:
	If mA = 3, mC = 9
	3/9 == 0.3 - more concrete elements than abstract ones.  Low abstraction.
	makes sense ✅

	If mA = 1, mC = 100
	1/100 == 0.01 - 1 abstract element per 100 concrete ones. Extremely low abstraction.
	makes sense ✅

	If mA = 0, mC = 100
	0/100 == 0 - 0 abstract elements. Abstraction is zero.
	makes sense ✅

	If mA = 100, mC = 100
	100/100 == 100 - For every abstract element there is one concrete element. Maximum abstraction.
	makes sense ✅

Controversial cases:

	If mA = 100, mC = 2
	100/2 == 50 - ??? Abstraction - 50? 
	The function result rightly returns 50, signaling abnormally high abstraction in the code.
	The function conditions do not say, that we can't pass more abstract that concrete elements to it. 

	If mA = 100, mC = 0
	100/0 == ??? 100 abstract elements to 0 concrete ones. Is this an abstraction == 1? Or should we still show the user the abnormally high abstractness?

So...
On the one hand, in the LCOM96b, Instability and DMS functions (but not in the regular LCOM) we return only values from 0 to 1 and it is logical to return the same here.

On the other hand, if we decide to manually return 1, then cases with maximum abstraction (3/3 == 1) and abnormal abstraction (100/1 == 1) will be equal.

How do we handle this?